EIPS/eip-7928.md

```diff
@@ -0,0 +1,171 @@
+            access_entry = {"tx_index": idx}
+            if is_write:
+                access_entry["value_after"] = value
+            
+            computed_access_list[(address, slot)].append(access_entry)
+        
+        # Process balance changes
+        for address, delta in balance_changes:
+            if address not in computed_balance_diffs:
+                computed_balance_diffs[address] = []
+            computed_balance_diffs[address].append({"tx_index": idx, "delta": delta})
+
+    # Validate access list and balance diffs
+    if not validate_access_list(block.block_access_list, computed_access_list):
+        raise InvalidBlock("Mismatch in block-level access list.")
+        
+    if not validate_balance_diffs(block.balance_diffs, computed_balance_diffs):
+        raise InvalidBlock("Mismatch in balance diffs.")
+```
+
+For contract deployments, the `value_after` must contain the runtime bytecode of the newly deployed contract.
+
+## Rationale
+
+### Full BAL Design Choice
+
+We chose the BAL variant for several key reasons:
+
+1. **Balance between size and parallelization benefits**: BALs enable both parallel disk reads and parallel EVM execution while keeping the size manageable.
+
+2. **Storage value inclusion**: Including post-execution values for write operations facilitates state reconstruction during syncing, enabling faster chain catch-up.
+
+3. **Balance diffs**: Including balance diffs ensures correct handling of execution that depends on account balances, crucial for parallel transaction execution.
+
+4. **Efficiency**: An analysis of historical blocks shows that Full BALs would average around 57 KiB with balance diffs adding only 9.6 KiB on average, which is a reasonable overhead given the performance benefits.
+
+### Block Size Considerations
+
+Including access lists increases block size, potentially impacting network propagation times. Based on analysis of historical blocks:
+
+- Average BAL size over 1,000 blocks was around 57 KiB (SSZ-encoded)
+- Average balance diff size was approximately 9.6 KiB
+- Worst-case BAL size for consuming the entire block gas limit with storage access operations would be approximately 0.93 MiB
+- Worst-case balance diff size would be around 0.11 MiB
+
+These sizes are manageable and less than the current worst-case block size achievable through calldata.
+
+### Asynchronous Verification
+
+Block execution can proceed with parallel IO and parallel EVM operations, with verification of the access list occurring alongside execution, ensuring correctness without delaying block processing.
+
+## Backwards Compatibility
+
+This proposal requires changes to the block structure that are not backwards compatible and require a hard fork.
+
+## Security Considerations
+
+### Validation Overhead
+
+Nodes must verify that the access lists and balance diffs in the block accurately reflect the actual accesses and balance changes during execution. Discrepancies MUST result in block invalidation.
```

Comment:
Worth mentioning the BAL should be canonical?
e.g., can't contain more info than needed.

Reply:
agree, will add!